### DISCUSSION TITLE
Making a proper deployment build for MacOS

### DISCUSSION THREAD

**Discussion by kheki4** [openPupil/Open-PupilEXT#57] [Q&A]
Dear Everyone,

While trying to make the deployment file for the Experimental-Community-Version release v0.1.2 we ran into a problem. The compiled .app file can be successfully placed inside a .dmg file to create an tiny installer-like file, that is fine. It can be unpacked and added to the Applications, and run with no problem on the very same machine that it was built on. However, on another Mac machine (which also does not have Xcode, Qt, or anything, only Pylon installed) it cannot be started after adding to the Applications. 

Our compiled .app file is ~24 Mb in size, which is circa 70 Mb smaller than the release v0.1.1. of PupilEXT, which is already a red flag. The former seems to lack the dependencies/libraries embedded within the .app file. I inspected it, and it does not at all contain the Frameworks folder, which would normally be there and likely contain a bunch of .dylib files, as opposedly is the case in the v0.1.1 release.

I found the MacOS_Deployment_Helper Matlab script in the Misc folder in the original source code already, however it seems to only change several existing item names/entries inside the resulting .app file in the Frameworks folder, but itself does not add them. 

I am really new to building for MacOS and do not clearly know what I should do to let CMake include the needed libraries in the .app file created. I already tried several hours of googling and looking around on SO, trying to make a working build, but I have not yet found a clear solution. (Also I only have one Mac so I need to wait for a friend to test it for me at each trial-error iteration, so that is the reason why I proceed with this matter so slowly.)

Could somebody please help or refer me to a page where a solution is likely to be found?
Thank you for your time!

**Comment by BZandi** [comment]
I noticed your issue with deploying the PupilEXT app on MacOS. Below is a step-by-step guide that should help you embed the necessary libraries and frameworks into your `.app` file.

### Step-by-Step Guide for Deploying PupilEXT on Mac

1. **Modify `CMakeLists.txt`**
    Make sure your `add_executable()` line in `CMakeLists.txt` looks like this:
    ```cma
    add_executable(PupilExt MACOSX_BUNDLE main.cpp ${RESOURCE_ADDED} mainwindow.cpp ...)
    ```

2. **Build the App**
    Build your project to create the `.app` file.

3. **Use `macdeployqt` Tool**
    Navigate to the build directory where your `.app` file is located. Run the following command to deploy Qt libraries:
    ```shell
    /path/to/Qt/5.15.0/clang_64/bin/macdeployqt PupilEXT.app
    ```

4. **Resolve Additional Dependencies**
    Note that for this step, I provided a Matlab script (MacOS_Deployment_Helper), allowing to link the dependencies automatically. However, it might be possible that more libs need to be linked, but the procedure is the same. 

    - **Install Pylon**: Ensure Pylon is installed on the target machine.
    - **Adjust Library Dependencies**:
        ```shell
        cd path/to/PupilEXT.app/Contents/MacOS
        otool -L PupilEXT

        install_name_tool -change /usr/local/Cellar/ffmpeg/4.3.2/lib/libswresample.3.dylib @executable_path/../Frameworks/libswresample.3.dylib PupilEXT
        install_name_tool -change /usr/local/Cellar/ffmpeg/4.3.2/lib/libavutil.56.dylib @executable_path/../Frameworks/libavutil.56.dylib PupilEXT
        install_name_tool -change /usr/local/Cellar/ffmpeg/4.3.2/lib/libavcodec.58.dylib @executable_path/../Frameworks/libavcodec.58.dylib PupilEXT
        install_name_tool -change /usr/local/Cellar/ffmpeg/4.3.2/lib/libavformat.58.dylib @executable_path/../Frameworks/libavformat.58.dylib PupilEXT
        install_name_tool -change /usr/local/Cellar/nettle/3.7.1/lib/libnettle.8.dylib @executable_path/../Frameworks/libnettle.8.dylib PupilEXT
        install_name_tool -change /usr/local/Cellar/ffmpeg/4.3.2/lib/libswscale.5.dylib @executable_path/../Frameworks/libswscale.5.dylib PupilEXT
        install_name_tool -change /usr/local/Cellar/ffmpeg/4.3.2/lib/libavresample.4.dylib @executable_path/../Frameworks/libavresample.4.dylib PupilEXT
        ```

5. **Create a DMG File**
    After fixing dependencies, you can package the app into a `.dmg` file:
    ```shell
    /path/to/Qt/5.15.0/clang_64/bin/macdeployqt PupilEXT.app -dmg
    ```

By following these steps, you should be able to create a deployable `.app` file that includes all necessary libraries and dependencies. I hope this helps.
